### PR TITLE
Add low-pT electrons to MINIAOD, update ID, improve end user experience (back port of 31220)

### PR DESCRIPTION
**Pull request by bainbrid**
#### PR description:

This PR relies on the following PR: https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/17 (UPDATED!)

- Adds low-pT electrons to MINIAOD by default
- Updates the ID ("2020Sept15" from BParking studies)
- Keeps only a subset of low-pT electrons (BDT score>1.5, mistag ~10%, efficiency ~90%) 
- Embed informations in the pat::Electron, and drops RECO collections
- Stores GBRForest object (for ID) in ROOT file format rather than XML, fixes [this](https://github.com/cms-sw/cmssw/issues/28780) issue
- Uses the `run2_miniAOD_UL` modifier to run PAT on existing RECO/AOD samples

#### PR validation:

- Tested locally with wfs () 
- Tested with a standalone "unit test" module. 
- Tested against PR #31220 to ensure consistent output.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

- This PR is a back port of cms-sw/cmssw#31080
- These developments will be used as part of the re-miniAOD UltraLegacy campaign. 

@slava77 @guitargeek @crovelli @afiqaize @SohamBhattacharya @jpata @jainshilpi @lsoffi

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test with cms-data/RecoEgamma-ElectronIdentification#15

**Comment by slava77** [comment]
> When I ran the RelVals I found an error in the following workflows:
> 135.4 step3

```
   [5] Calling method for module PATElectronProducer/'patLowPtElectrons'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: edm::ValueMap<float>
Looking for module label: rekeyLowPtGsfElectronSeedValueMaps
Looking for productInstanceName: ptbiased
```

looks related.
Something is not (un)setup right for the fastsim case.

**Comment by bainbrid** [comment]
In respose to https://github.com/cms-sw/cmssw/pull/32372#issuecomment-737619284.

The FastSim-related config is unchanged. 

In 11_X, the `rekeyLowPtGsfElectronSeedValueMaps` collection is present in the step1.root file. 
While for 10_6_X it is not. Hence the failed step3 test. 

After some digging, I realised that the structure of RecoEgamma_EventContent_cff.py has changed. In 11_X, I added the rekeyLowPtGsfElectronSeedValueMaps collection to the RecoEgammaAOD list, which is then cloned and extended by RecoEgammaRECO and RecoEgammaFEVT. In 10_6_X, the latter two are defined independently of RecoEgammaAOD and my back port didn't account for this. 

Now corrected with the latest commit. Tested locally with wf 135.4.

**Comment by slava77** [comment]
test parameters:
- pull_request = cms-data/RecoEgamma-ElectronIdentification#15
- workflows = 136.898,1304.181

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@smuzaffar 
I'm confused about the tests status labels.
From the messages above, I think that the tests are done and the comparisons are available in https://github.com/cms-sw/cmssw/pull/32372#issuecomment-738183247
Please check why the labels are still in "tests-started"
Thank you.

**Comment by smuzaffar** [comment]
@slava77 , I think the test here were started when I updated the bot to use commit statuses for updating label. It missed a stats "cms/arch/required" and that is why bot did not set the label. I have force to set the needed status and label is properly set now.

**Comment by bainbrid** [comment]
@cavalari

**Comment by slava77** [comment]
> Reco comparison results: 28 differences found in the comparisons

in many workflows we see a difference in `lowPtGsfElectronID`.
This is not allowed in 10_6_X. Only changes in the UL re-miniAOD workflow are allowed (136.88811 in the jenkins tests).
Please make updates so that only `run2_miniAOD_UL` is picking up the new ID.

**Comment by slava77** [comment]
> @slava77 , I think the test here were started when I updated the bot to use commit statuses for updating label. It missed a stats "cms/arch/required" and that is why bot did not set the label. I have force to set the needed status and label is properly set now.

@smuzaffar 
I see that  `tests-approved` label was added, this is good.
What about `comparison-available`? is it still expected to be assigned?

**Comment by smuzaffar** [comment]
I have dropped the comparison approved label. It does not make any sense. State of comparisons are now part of tests, so once tests approved label is set then that means the comparison has been run too.

**Comment by smuzaffar** [comment]
by the way, soon I will merge the test and comparison results in one comment ( just to avoid an extra github email notification)

**Comment by bainbrid** [comment]
@slava77 in response to https://github.com/cms-sw/cmssw/pull/32372#issuecomment-739119758, I'm just doing final checks. 

It took me quite some time yesterday and today to figure out the version of the original XML that I need to maintain the status quo in 10_6_X RECO, but I've got there. (Just final checks now.)

My issue is that the XML file has since moved on, so I need to create a new branch with the original XML and the "new" 2020Sept15 model and make a PR for 10_6_X.
In short, the new PR will have:
- the original XML file used throughout the 10_6_X cycle
- the 2020Sept15 model, unchanged w.r.t. 11_3_X, for reminiAOD

**Comment by bainbrid** [comment]
@slava77 the latest commit:
- reverts to the original XML model used throughout the 10_6_X cycle 
- uses the 2020Sept15 model for the UL reminiAOD

**This PR now relies on the following PR: cms-data/RecoEgamma-ElectronIdentification#17**

**Comment by bainbrid** [comment]
These changes https://github.com/cms-sw/cmssw/pull/32372#issuecomment-739540773 tested locally on 25202.0 (TTbar), 1325.518 (TTbar reminiAOD), 136.88811 (JetHT, 2018D, reminiAOD).

**Comment by slava77** [comment]
> **This PR now relies on the following PR: [cms-data/RecoEgamma-ElectronIdentification#17](https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/17)**

I don't really understand why this dependence; it looks the same as cms-data/RecoEgamma-ElectronIdentification#15

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@smuzaffar 
the label status is still not updating here appropriately.
I think that `tests-started` is incorrect, because tests were complete some 15 hours ago.

**Comment by smuzaffar** [comment]
ah right, the required test flag was missing for old release cycles. This is fixed now ( https://github.com/cms-sw/cms-bot/commit/375f0077f653c61e6aa0d5b39a7110955fe76120)

**Comment by slava77** [comment]
test parameters:
- pull_request = cms-data/RecoEgamma-ElectronIdentification#17
- workflows = 136.898,1304.181

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
ehm, now the bot is not reacting to my test requests

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> ehm, now the bot is not reacting to my test requests

actually, it looks like the request was made (# 1​14​05  in https://cmssdt.cern.ch/jenkins/job/ib-run-pr-tests/#, but the infrastructure is busy; I'd assume that an identical request is not redone)

File: CommonTools/MVAUtils/src/GBRForestTools.cc [hunk 1]
```diff
@@ -117,6 +119,14 @@ namespace {
   }
 
   std::unique_ptr<GBRForest> init(const std::string& weightsFileFullPath, std::vector<std::string>& varNames) {
+    //
+    // Load weights file, for ROOT file
+    //
+    if (reco::details::hasEnding(weightsFileFullPath, ".root")) {
+      TFile gbrForestFile(weightsFileFullPath.c_str());
+      return std::unique_ptr<GBRForest>(reinterpret_cast<GBRForest*>(gbrForestFile.Get("gbrForest")));
```
[the remark this answers is not in the document]
Comment (bainbrid) [reply CommonTools/MVAUtils/src/GBRForestTools.cc]: Done.

File: PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py [hunk 2]
```diff
@@ -11,42 +11,49 @@
...
+    # MC matching
+    genParticleMatch = "lowPtElectronMatch",
+
+    # Electron ID
+    addElectronID = True,
+    electronIDSources = dict(
+        unbiased = cms.InputTag("rekeyLowPtGsfElectronSeedValueMaps:unbiased"),
+        ptbiased = cms.InputTag("rekeyLowPtGsfElectronSeedValueMaps:ptbiased"),
+        ID       = cms.InputTag("lowPtGsfElectronID"),
+    ),
```
[the remark this answers is not in the document]
Comment (bainbrid) [reply PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py:27]: Some comments from me below ...

From a RECO perspective, this PR won't change the bParking content.

From a miniAOD perspective, this PR won't change the bParking content, but currently it will change how this content is accessed. 

Specifically, in the current HEAD of 10_6_X, the slimmedLowPtElectrons are stored along with several associated RECO collections (lowPtGsfElectronCores, lowPtGsfTracks, etc...). This PR will leave the content of those collections unchanged, but it will deprecate these collections for miniAOD and instead embed their content directly within the pat::Electron.

If there is a reprocessing of the BParking data sets in 10_6_X, then it is my opinion that the bParking era should continue to be used.

And here are some further comparative comments on the anticipated changes to 10_6_X due to this PR.

In the current HEAD of CMSSW_10_6_X, the bParking modifier is used in:
- [lowPtGsfElectronSeeds_cfi](https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_X/RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectronSeeds_cfi.py#L50-L53) to lower the pT threshold to 0.5 GeV and loosen the BDT threshold to "Very Loose",
- [MicroEventContent_cff](https://github.com/cms-sw/cmssw/pull/32372/files) to add the slimmedLowPtElectrons and associated RECO collections to miniAOD.

In this PR:
- for [lowPtGsfElectronSeeds_cfi](https://github.com/bainbrid/cmssw/blob/LowPtElectrons_userFloats_106X/RecoEgamma/EgammaElectronProducers/python/lowPtGsfElectronSeeds_cfi.py#L50-L53), the same bParking modifier logic will be applied (0.5 GeV and "Very Loose"),
- for [MicroEventContent_cff](https://github.com/bainbrid/cmssw/blob/LowPtElectrons_userFloats_106X/PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py#L92-L94), the slimmedLowPtElectrons (and the associated gsfTracksOpenConversions RECO collection) will  now be saved to miniAOD _by default_. Also, as stated above, the information from the deprecated RECO collections will be embedded in the low-pT pat::Electron.
- in [lowPtElectronSelector_cfi.py](https://github.com/bainbrid/cmssw/blob/LowPtElectrons_userFloats_106X/PhysicsTools/PatAlgos/python/selectionLayer1/lowPtElectronSelector_cfi.py#L12-L14), the bParking modifier is used to ensure no selection is applied (to be consistent with the bParking miniAOD content in 10_2_X)

Hopefully this is helpful ...

File: PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py [hunk 3]
```diff
@@ -61,3 +68,28 @@
                                                            electronSource = "gedGsfElectrons",
                                                            genParticleMatch = "electronMatch"
                                                            )
+
+# Schedule rekeying of seed BDT ValueMaps by reco::GsfElectron for run2_miniAOD_UL
+from Configuration.ProcessModifiers.run2_miniAOD_UL_cff import run2_miniAOD_UL
+from RecoEgamma.EgammaElectronProducers.lowPtGsfElectronSeedValueMaps_cff import rekeyLowPtGsfElectronSeedValueMaps
+from RecoEgamma.EgammaElectronProducers.defaultLowPtGsfElectronID_cfi import defaultLowPtGsfElectronID
```
[the remark this answers is not in the document]
Comment (bainbrid) [reply PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py]: Done. I could add these changes to https://github.com/cms-sw/cmssw/pull/32391?

File: PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py [hunk 4]
```diff
@@ -61,3 +68,28 @@
...
+    Version = "V1",
+)
+run2_miniAOD_UL.toModify(
+    patLowPtElectrons,
+    electronIDSources = dict(
+        unbiased = cms.InputTag("rekeyLowPtGsfElectronSeedValueMaps:unbiased"),
+        ptbiased = cms.InputTag("rekeyLowPtGsfElectronSeedValueMaps:ptbiased"),
+        ID       = cms.InputTag("defaultLowPtGsfElectronID"),
+    )
+)
```
[the remark this answers is not in the document]
Comment (bainbrid) [reply PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py]: Done.

File: PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py [hunk 5]
```diff
@@ -61,3 +68,28 @@
...
+_makePatLowPtElectronsTask = makePatLowPtElectronsTask.copy()
+_makePatLowPtElectronsTask.add(rekeyLowPtGsfElectronSeedValueMaps)
+_makePatLowPtElectronsTask.add(defaultLowPtGsfElectronID)
```
[the remark this answers is not in the document]
Comment (bainbrid) [reply PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py]: Done.

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc [hunk 6]
```diff
@@ -1,84 +1,187 @@
...
+#include "DataFormats/Common/interface/View.h"
 #include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
 #include "FWCore/Utilities/interface/InputTag.h"
-#include "RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "CommonTools/MVAUtils/interface/GBRForestTools.h"
+#include "DataFormats/EgammaCandidates/interface/GsfElectronFwd.h"
```
[the remark this answers is not in the document]
Comment (bainbrid) [reply RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc:13]: Agree.

**Comment by bainbrid** [comment]
I forced push as I had to rebase to CMSSW_10_6_X_2020-12-02-2300 due to an expired IB.

These two new commits are in response to @slava77 's feedback yesterday: https://github.com/cms-sw/cmssw/pull/32372/commits/40916e680d00adf2f9489e09df4dba29325067e9 and https://github.com/cms-sw/cmssw/pull/32372/commits/ee6ceb36b6b166d07d876ac9b27c12749a6ba3eb

File: PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rappoccio) [reply PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py:27]: > @silviodonato @rappoccio please clarify/confirm if we need to consider possible `bParking` production in the 10_6_X context

There are currently no official requests for UL of the bParking. That being said, it is still an option that may happen in early 2021.

**Comment by silviodonato** [comment]
test parameters:
- pull_request = cms-data/RecoEgamma-ElectronIdentification#15
- workflows = 136.898,1304.181
- addpkg = TauAnalysis/MCEmbeddingTools

**Comment by silviodonato** [comment]
please test
(you might need to backport #32400)

**Comment by slava77** [comment]
> ROOTFileChecks: Some differences in event products or their sizes found

```
Change in products found in 10224.0_TTbar_13+*/step3_inMINIAODSIM.root
    Product added patElectrons_slimmedLowPtElectrons__RECO.
    Product added recoConversions_gsfTracksOpenConversions_gsfTracksOpenConversions_RECO.
Change in products found in 10224.0_TTbar_13+*step3.root
    Product added floatedmValueMap_rekeyLowPtGsfElectronSeedValueMaps_ptbiased_RECO.
    Product added floatedmValueMap_rekeyLowPtGsfElectronSeedValueMaps_unbiased_RECO.
```

so, this means that the default production workflows are running the new low-pt modules, while normally for the no-change policy we should have it running only with the `run2_miniAOD_UL`.
perhaps an easier approach is to add the products above only with this modifier, although making the additions to the relevant tasks would be better.

**Comment by slava77** [comment]
> pull_request = [cms-data/RecoEgamma-ElectronIdentification#15](https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/15)

this is not the right external.
@silviodonato was this override of existing parameter intentional?

**Comment by slava77** [comment]
@cmsbuild please abort test

**Comment by slava77** [comment]
> > pull_request = [cms-data/RecoEgamma-ElectronIdentification#15](https://github.com/cms-data/RecoEgamma-ElectronIdentification/pull/15)
> 
> this is not the right external.
> @silviodonato was this override of existing parameter intentional?

what I think was the right settings were made above in https://github.com/cms-sw/cmssw/pull/32372#issuecomment-740104326

**Comment by silviodonato** [comment]
please abort

**Comment by silviodonato** [comment]
test parameters:
- pull_request = cms-data/RecoEgamma-ElectronIdentification#17
- workflows = 136.898,1304.181
- addpkg = TauAnalysis/MCEmbeddingTools

**Comment by bainbrid** [comment]
> > ROOTFileChecks: Some differences in event products or their sizes found
> 
> ```
> Change in products found in 10224.0_TTbar_13+*/step3_inMINIAODSIM.root
>     Product added patElectrons_slimmedLowPtElectrons__RECO.
>     Product added recoConversions_gsfTracksOpenConversions_gsfTracksOpenConversions_RECO.
> Change in products found in 10224.0_TTbar_13+*step3.root
>     Product added floatedmValueMap_rekeyLowPtGsfElectronSeedValueMaps_ptbiased_RECO.
>     Product added floatedmValueMap_rekeyLowPtGsfElectronSeedValueMaps_unbiased_RECO.
> ```
> 
> so, this means that the default production workflows are running the new low-pt modules, while normally for the no-change policy we should have it running only with the `run2_miniAOD_UL`.
> perhaps an easier approach is to add the products above only with this modifier, although making the additions to the relevant tasks would be better.

Apologies, an oversight on my part. I need remove the two lines added in [this](https://github.com/cms-sw/cmssw/pull/32372/files?file-filters%5B%5D=.py#diff-65a9388b6207e2d4ef15fdc13d9bbeee170bc88cc5496f3a1a560f9cf977c09f) commit. (Testing now...)

**Comment by bainbrid** [comment]
Actually, reverting [this](https://github.com/cms-sw/cmssw/pull/32372/files?file-filters%5B%5D=.py#diff-65a9388b6207e2d4ef15fdc13d9bbeee170bc88cc5496f3a1a560f9cf977c09f) change means that the collection does not exist for the PAT sequence unless the run2_miniAOD_UL is specified, as done [here](https://github.com/cms-sw/cmssw/pull/32372/files?file-filters%5B%5D=.py#diff-301347daec1998142b2c5167fcec9c40bacb33d2622ae294648e9be73a34ee40R77). However, the run2_miniAOD_UL modifier is not for this wf. How should I proceed? Should I add it to the PAT sequence regardless of whether the run2_miniAOD_UL is specified or not?

**Comment by slava77** [comment]
> However, the run2_miniAOD_UL modifier is not for this wf.

"this" is an unresolved pointer. Please clarify which workflow do you have in mind.

The minimal setup would have only the workflows with `run2_miniAOD_UL` modified by this PR.

**Comment by bainbrid** [comment]
this = 10224.0.

So perhaps a simple approach is to only extend the MicroEventContent to include low pT electrons for run2_miniAOD_UL (and bParking)?

**Comment by slava77** [comment]
> this = 10224.0.
> 

got it (so, my comments stand, this workflow ideally should not be modified).

> So perhaps a simple approach is to only extend the MicroEventContent to include low pT electrons for run2_miniAOD_UL (and bParking)?

this should work as well, although it's a bit less ideal because the modules will still be added to the tasks.

**Comment by bainbrid** [comment]
I propose two changes:

1) only adding low pT to MINIAOD for run2_miniAOD_UL and bParking

```
+from Configuration.ProcessModifiers.run2_miniAOD_UL_cff import run2_miniAOD_UL
 from Configuration.Eras.Modifier_bParking_cff import bParking
-_bParking_extraCommands = ['keep *_slimmedLowPtElectrons_*_*',
-                           'keep recoGsfElectronCores_lowPtGsfElectronCores_*_*',
-                           'keep recoSuperClusters_lowPtGsfElectronSuperClusters_*_*',
-                           'keep recoCaloClusters_lowPtGsfElectronSuperClusters_*_*',
-                           'keep recoGsfTracks_lowPtGsfEleGsfTracks_*_*',
-                           'keep floatedmValueMap_lowPtGsfElectronSeedValueMaps_*_*',
-                           'keep floatedmValueMap_lowPtGsfElectronID_*_*',
-                           'keep *_lowPtGsfLinks_*_*',
-                           'keep *_gsfTracksOpenConversions_*_*',
-                           ]
-bParking.toModify(MicroEventContent, outputCommands = MicroEventContent.outputCommands + _bParking_extraCommands)
+_lowPt_extraCommands = ['keep *_slimmedLowPtElectrons_*_*',
+                        'keep *_gsfTracksOpenConversions_*_*',]
+(bParking | run2_miniAOD_UL).toModify(MicroEventContent,
+                                      outputCommands = MicroEventContent.outputCommands + _lowPt_extraCommands)
```

2) reverting [this](https://github.com/cms-sw/cmssw/pull/32372/files?file-filters%5B%5D=.py#diff-65a9388b6207e2d4ef15fdc13d9bbeee170bc88cc5496f3a1a560f9cf977c09f) change so that the RECO sequence remains unchanged. 

Satisfactory?

**Comment by slava77** [comment]
> I propose two changes:
> 
[...]
> 
> Satisfactory?


at a glance, both look fine.
Thank you.

**Comment by bainbrid** [comment]
Shall I add to this back port the changes here? https://github.com/cms-sw/cmssw/pull/32400

**Comment by slava77** [comment]
> Shall I add to this back port the changes here? #32400

already suggested in https://github.com/cms-sw/cmssw/pull/32372#issuecomment-740617745

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by bainbrid** [comment]
These commits (https://github.com/cms-sw/cmssw/pull/32372/commits/c1513e275eca1bd8a919fc448d6e8b37cb582b2e and https://github.com/cms-sw/cmssw/pull/32372/commits/d5297a2b3adc431ae72a0c8bc3c015f40d920424) ensure:
- the RECO sequence is unchanged in 10_6_X, 
- low pT electrons are added to MINIAOD _only_ for bParking and run2_miniAOD_UL.

Tested locally with wfs 10224.0 and 1325.518.

The commit https://github.com/cms-sw/cmssw/pull/32372/commits/692e790578ad6e1682d43974fe2bfd4dffffe56c fixes the failed TauAnalysis/MCEmbeddingTools unit test (and is the back port of #32400). Tested locally by running `TauAnalysis/MCEmbeddingTools/test/runtests.sh`.

**Comment by slava77** [comment]
the bParking miniAOD is apparently broken 
in 1304.181

```
   [5] Calling method for module PATElectronProducer/'patLowPtElectrons'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: edm::ValueMap<float>
Looking for module label: rekeyLowPtGsfElectronSeedValueMaps
Looking for productInstanceName: ptbiased
```

**Comment by bainbrid** [comment]
Was looking into it. Already tested on wf 1304.181. Now also testing 136.898.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (slava77) [reply RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc:13]: there was no update following this "agree" comment.
What is the status/plan?

[the remark this answers is not in the document]
Comment (bainbrid) [reply RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc:13]: My understanding is that this part of the back port should remain "as is". i.e. no changes.

File: PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (slava77) [reply PhysicsTools/PatAlgos/python/producersLayer1/lowPtElectronProducer_cff.py:27]: > From a miniAOD perspective, this PR won't change the bParking content, but currently it will change how this content is accessed.

apparently it is not the case: this PR is adding electronID variables to the lowPtElectrons running with the `bParking` modifier, as can be seen in the outputs of wf 1304.181 and 136.898
This change is triggered by the lines highlighted in this comment subthread (`addElectronID = True ` ... )

Since only new variables are saved in the products and apparently these variables were already computed for other reasons, I think that it should be acceptable to keep this change.

File: RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (slava77) [reply RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc:13]: > My understanding is that this part of the back port should remain "as is". i.e. no changes.

this disagrees with the earlier comment to the code

> it looks like this part of the backport better be still applied.

was meant to say that this part of the master should be also be backported.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/32372	541fb89 
- code changes are in line with the PR description and the follow up review; see the notes below on the congruence of this PR with the master
- jenkins (and some local) tests pass and comparisons with the baseline show expected/acceptable differences
    - `slimmedLowPtElectrons` now shows up in the UL re-miniAOD workflow 136.88811; this is the primary target of this PR
    - the bParking workflows (1304.181 and 136.898) have a change in `electronIDs` variables in `slimmedLowPtElectrons`; this is somewhat acceptable https://github.com/cms-sw/cmssw/pull/32372#discussion_r538884752 


this PR includes partial or complete backports of the following:
- #32400 : complete (TauAnalysis/MCEmbeddingTools/python/customisers.py)
- #31220 (the main source of this PR) : close to complete
    - using edm::View in RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronIDProducer.cc  was not backported
- #31080 (PR preceding #31220) is partial: changes in LowPtGsfElectronFeatures.{h,cc} , what's not superseded in #31220, are not backported, related mainly to the method(s) in `namespace lowptgsfeleseed {`
- #27169 and #27344 (clang-format PRs): only modifications in files relevant to #31220 are here
- #27477 , #28746, #31700 (various egamma refactoring PRs) contributed to LowPtGsfElectronSeedValueMapsProducer.{h,cc}

Small discrepancies are present here compared to the master:
- somewhat trivial differences in imported files in PhysicsTools/PatAlgos/python/slimming/slimmedLowPtElectrons_cff.py   is rectified in #32426 
- additional 106X specific updates were made to maintain the no-change policy; this does not need a sync with the master

**Comment by qliphy** [comment]
@cms-sw/simulation-l2 @cms-sw/analysis-l2   Would you please have a look?  This is urgently needed for re-MiniAOD.

**Comment by qliphy** [comment]
urgent

**Comment by civanch** [comment]
+1

**Comment by silviodonato** [comment]
merge

**Comment by santocch** [comment]
+1
